kentikapi/internal/synthetics/test.go

```diff
@@ -0,0 +1,652 @@
+		HTTPLatencyCriticalStdDev: convert.MillisecondsF32ToDuration(hs.HttpLatencyCriticalStddev),
+		HTTPLatencyWarningStdDev:  convert.MillisecondsF32ToDuration(hs.HttpLatencyWarningStddev),
+		HTTPValidCodes:            hs.HttpValidCodes,
+		DNSValidCodes:             hs.DnsValidCodes,
+		UnhealthySubtestThreshold: hs.UnhealthySubtestThreshold,
+		AlarmActivation:           alarmActivation,
+	}, nil
+}
+
+func pingSettingsFromPayload(ps *syntheticspb.TestPingSettings) *synthetics.PingSettings {
+	if ps == nil {
+		return nil
+	}
+
+	return &synthetics.PingSettings{
+		Timeout:  time.Duration(ps.Timeout) * time.Millisecond,
+		Count:    ps.Count,
+		Delay:    convert.MillisecondsF32ToDuration(ps.Delay),
+		Protocol: synthetics.PingProtocol(ps.Protocol),
+		Port:     ps.Port,
+	}
+}
+
+func tracerouteSettingsFromPayload(ts *syntheticspb.TestTraceSettings) *synthetics.TracerouteSettings {
+	if ts == nil {
+		return nil
+	}
+
+	return &synthetics.TracerouteSettings{
+		Timeout:  time.Duration(ts.Timeout) * time.Millisecond,
+		Count:    ts.Count,
+		Delay:    convert.MillisecondsF32ToDuration(ts.Delay),
+		Protocol: synthetics.TracerouteProtocol(ts.Protocol),
+		Port:     ts.Port,
+		Limit:    ts.Limit,
+	}
+}
+
+func alarmActivationFromPayload(as *syntheticspb.ActivationSettings) (*synthetics.AlarmActivationSettings, error) {
+	if as == nil {
+		return nil, nil
+	}
+
+	timeWindow, err := time.ParseDuration(as.TimeWindow + as.TimeUnit)
+	if err != nil {
+		return nil, fmt.Errorf("parse alarm activation time window %q: %v", as.TimeWindow+as.TimeUnit, err)
+	}
+
+	times, err := strconv.ParseUint(as.Times, 10, 0)
+	if err != nil {
+		return nil, fmt.Errorf("parse alarm activation times %q: %v", as.Times, err)
+	}
+
+	gracePeriod, err := strconv.ParseUint(as.GracePeriod, 10, 0)
+	if err != nil {
+		return nil, fmt.Errorf("parse alarm activation grace period %q: %v", as.GracePeriod, err)
+	}
+
+	return &synthetics.AlarmActivationSettings{
+		TimeWindow:  timeWindow,
+		Times:       uint(times),
+		GracePeriod: uint(gracePeriod),
+	}, nil
+}
+
+func userInfoPtrFromPayload(ui *syntheticspb.UserInfo) *synthetics.UserInfo {
+	if ui == nil {
+		return nil
+	}
+	return &synthetics.UserInfo{
+		ID:       ui.Id,
+		Email:    ui.Email,
+		FullName: ui.FullName,
+	}
+}
+
+// testToPayload converts synthetics test from model to payload. It sets only ID and read-write fields.
+func testToPayload(t *synthetics.Test) (*syntheticspb.Test, error) {
+	if t == nil {
+		return nil, errors.New("test object is nil")
+	}
+
+	ts, err := testSettingsToPayload(t.Settings, t.Type)
+	if err != nil {
+		return nil, err
+	}
+
+	return &syntheticspb.Test{
+		Id:            t.ID,
+		Name:          t.Name,
+		Type:          string(t.Type),
+		Status:        syntheticspb.TestStatus(syntheticspb.TestStatus_value[string(t.Status)]),
+		Settings:      ts,
+		Cdate:         nil, // read-only
+		Edate:         nil, // read-only
```

Comment:
We need to support the ability to provide `edate` in `TestUpdate` requests.  If provided, the API returns error if the object to be modified has been modified in the DB after the `edate` timestamp. This allows to guard against concurrent modifications.